Prioritize uninherited effect points for the omit barline effect

sw1tchbl4d3r opened the pull request:
Closes #22551

In stable, inherited points cannot set the omit barline effect, the checkbox is grayed out.
Usually, inherited points would inherit the omit barline effect setting from its uninherited parent, but this is not always the case, as with [the map in the aforementioned issue](https://osu.ppy.sh/beatmapsets/1770207#taiko/3625974)

It contains these two lines
```
11992,228.136882129278,4,1,0,90,1,8
11992,-136.979166666667,4,1,0,90,0,0
```

in which the first uninherited point sets omit barline to true (fourth bit set, `8`), while the inherited point after doesn't share the same effect (no bit set, `0`).
It's not clear to me how this happened, but since it's a ranked map it's probably best to support this behavior.
To support it, we have to prioritize the omit barline setting from the original timing point, so this patch will only set the omit barline effect for the last effect point it finds (which is the first in the pending list => from the original uninherited point)

Since this doesn't adress that inherited points shouldn't have the ability to switch this effect at all, I'm open for other ideas of implementing this.
There's two edge-cases we have to worry about:
1) There is no uninherited line, only an inherited line at a different point in time setting the effect: This will not affect barline generation here as the generator loops over all timing points and the effect points closest to them, which will always be the points generated with the timing point.
2) There are multiple uninherited lines at one precise time: If this were to happen, the setting of the last timing point in the file would be set here, which seems like desirable behavior, even though this is a very unlikely edgecase anyway.

sw1tchbl4d3r committed:
Prioritise timing points for the omit barline effect

peppy commented on osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:
```diff
@@ -485,16 +485,32 @@ private void addControlPoint(double time, ControlPoint point, bool timingChange)
         private void flushPendingPoints()
         {
             // Changes from non-timing-points are added to the end of the list (see addControlPoint()) and should override any changes from timing-points (added to the start of the list).
+            // As non-timing-points should not set the OmitFirstBarline setting, we will honor the setting of the last effect point, the effect point generated by the timing-point (if any).
```
This doesn't hold true in lazer, where you can actually set the setting on a non-timing point. I honestly don't know how valid that is, but it's a thing..

sw1tchbl4d3r replied:
The todo is specifically related to the stable beatmap format points that get fed into this function in the first place, from which we should honor the ones first in the list i.e. last in the loop more because they're effect points originating from timing points. Fixing that non timing points shouldnt be able to set this anyway might be an issue for another day.

peppy replied:
For sure. But as you had it this also would break lazer editor behaviour. Which can be seen by the first case in your tests (which asserts `false` but should probably assert `true`).

peppy commented on osu.Game.Tests/Resources/overlapping-control-points.osu:
```diff
@@ -4,16 +4,16 @@ osu file format v14
+0,-66.6666666666667,4,3,0,100,0,9
 
 // Inherited then timing (equivalent to previous)
 1000,-66.6666666666667,4,3,0,100,0,1
 1000,500,4,2,0,100,1,0
 
 // Inherited then timing (different to previous)
-2000,-133.333333333333,4,1,0,100,0,0
-2000,250,4,2,0,100,1,0
+2000,-133.333333333333,4,1,0,100,0,8
+2000,250,4,2,0,100,1,8
 
 // Timing then inherited (different to previous)
-3000,500,4,2,0,100,1,0
+3000,500,4,2,0,100,1,8
```
I don't think this test should have been changed like this. It's actually hiding a potential flaw in implementation, and makes it very hard to review / logically understand. Best to define a new `.osu` file for new cases like this.

peppy commented on osu.Game.Tests/Beatmaps/Formats/LegacyBeatmapDecoderTest.cs:
```diff
@@ -261,6 +261,11 @@ public void TestDecodeOverlappingTimingPoints()
                 Assert.That(controlPoints.EffectPointAt(2500).KiaiMode, Is.False);
                 Assert.That(controlPoints.EffectPointAt(3500).KiaiMode, Is.True);
 
+                Assert.That(controlPoints.EffectPointAt(500).OmitFirstBarLine, Is.False);
```
I believe this would be wrong for the lazer editor (where the order specified in your test *can* actually exist).

peppy commented on osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:
```diff
@@ -485,16 +485,32 @@ private void addControlPoint(double time, ControlPoint point, bool timingChange)
+                    else
+                        effectPoint.OmitFirstBarLine = pendingEffectPoint.OmitFirstBarLine;
+                }
+                else
```
Well okay, maybe it still works because of your manual tracking via the `effectPoint` variable. But I believe the whole method would read better if you remove the reliance on that variable.

I think a better way of writing this would be:

```csharp
        private void flushPendingPoints()
        {
            // Changes from non-timing-points are added to the end of the list (see addControlPoint()) and should override any changes from timing-points (added to the start of the list).

            for (int i = pendingControlPoints.Count - 1; i >= 0; i--)
            {
                var cp = pendingControlPoints[i];
                var type = cp.GetType();

                if (pendingControlPointTypes.Contains(type))
                {
                    // In stable, OmitFirstBarLine could only be set on timing points (ie. non-inherited points).
                    // From this perspective, it seems odd that we have the OmitFirstBarLine setting in EffectControlPoint as this breaks that assumption.
                    //
                    // To make sure legacy beatmaps are handled correctly, we therefore need to make sure to respect the earliest OmitFirstBarLine setting
                    // for the current point in time (aka the one read first from the file).
                    //
                    // The easiest way to handle this is to transfer any non-false values from the current flush group to the actual ControlPointInfo instance.
                    if ((cp as EffectControlPoint)?.OmitFirstBarLine == true)
                    {
                        beatmap.ControlPointInfo.EffectPointAt(cp.Time).OmitFirstBarLine = true;
                    }

                    continue;
                }

                pendingControlPointTypes.Add(type);
                beatmap.ControlPointInfo.Add(pendingControlPointsTime, cp);
            }

            pendingControlPoints.Clear();
            pendingControlPointTypes.Clear();
        }

```

If you'd like to attempt to fix the tests and apply this method then I'll leave it to you, or if you'd prefer I can open a second PR with this and accompanying tests.

sw1tchbl4d3r replied:
I'll do so later today.

sw1tchbl4d3r committed:
Allow OmitBarLine effect from any point

sw1tchbl4d3r committed:
capitalization

sw1tchbl4d3r commented:
The proposed idea needed a few more touch ups in regards to redundant effect points, the comments should explain it, other than that seems like the better approach.
We could also force add effect points like we do with timing points, but that seems less pretty.

peppy commented on osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:
```diff
@@ -485,31 +485,55 @@ private void addControlPoint(double time, ControlPoint point, bool timingChange)
+                    {
+                        var prevPoint = beatmap.ControlPointInfo.EffectPointAt(pendingControlPointsTime);
+
+                        // If this point is the default point, or does not share our points time, that means that the ControlPointInfo.Add() call
+                        // for the pendingPoint below did not actually add a point.
+                        //
+                        // This can happen if the point is deemed redundant due to it not changing effects.
+                        // If this happens, we need to manually re-add an effect point with OmitFirstBarLine now set to true, inheriting the rest of the settings
+                        // from the previous effect point
```
This reads really bad. Try using the return value from `ControlPointInfo.Add` to only add to the pending list if a point was actually created.

sw1tchbl4d3r replied:
The problem with this is that we need the value on the next iteration. We could refrain from adding to the types list if the add failed, but then we might add a second point that normally would not have been added, breaking tests. The only way to do it this way is to get yet another variable to store the last iterations adding result, is that the way here?

peppy replied:
I think you have things in a pretty... best-they-can-be state. I tried a few other simplified approaches but they all have flaws in one way or another.

I do wonder if just moving this property to `TimingControlPoint` is lower effort at this point, but oh well.

sw1tchbl4d3r committed:
Improve readability

peppy committed:
Remove `DeepClone` as unnecessary and adjust comments a bit

peppy committed:
Add back `DeepClone`

Turns out this will be required (even though there's no test coverage
proving it) in the case that the previous control point was equivalent
but non-default.

Without this copy, we could lose information.

peppy commented:
This is going to require a "quick" check by someone else unfortunately.
